scripts/bootloader/provision.py

```diff
@@ -13,45 +13,87 @@
 from hashlib import sha256
 import os
 
-# Size of LCS storage in OTP in bytes
+# Size of LCS storage and implementation ID in OTP in bytes
 LCS_STATE_SIZE = 0x8
 IMPLEMENTATION_ID_SIZE = 0x20
 NUM_BYTES_PROVISIONED_ELSEWHERE = LCS_STATE_SIZE + IMPLEMENTATION_ID_SIZE
 
-def generate_provision_hex_file(s0_address, s1_address, hashes, provision_address, output, max_size,
-                                num_counter_slots_version):
-    # Add addresses
-    provision_data = struct.pack('III', s0_address, s1_address,
-                                 len(hashes))
+def generate_provision_intel_hex_file(provision_data, provision_address, output, max_size):
+    assert len(provision_data) <= max_size, """Provisioning data doesn't fit.
+Reduce the number of public keys or counter slots and try again."""
 
-    for mhash in hashes:
-        provision_data += struct.pack('I', 0xFFFFFFFF) # Invalidation token
-        provision_data += mhash
+    ih = IntelHex()
+    ih.frombytes(provision_data, offset=provision_address)
+    ih.write_hex_file(output)
 
-    num_counters = 1 if num_counter_slots_version > 0 else 0
-    provision_data += struct.pack('H', 1) # Type "counter collection"
-    provision_data += struct.pack('H', num_counters)
+
+def add_hw_counters(provision_data, num_counter_slots_version, mcuboot_counters_slots):
+    num_counters = min(mcuboot_counters_slots, 1) + min(num_counter_slots_version, 1)
+
+    if num_counters == 0:
+        return provision_data
 
     assert num_counter_slots_version % 2 == 0, "--num-counters-slots-version must be an even number"
+    assert mcuboot_counters_slots    % 2 == 0, "--mcuboot-counters-slots     must be an even number"
```

Comment:
I'm just curious where this even slot number comes from?

On another side note, does this also consider the slot used in NSIB? So if I have NSIB+MCUBoot, it would currently be 2, but if we start supporting multi-image counters in MCUBoot, we would have 3?

Reply:
provision.py used to round up the number of slots to the next even number, giving you a different number of configured and actual slots. I rectified that recently.

I don't know why it did this.

There must be some piece of infrastructure that works on words.

Maybe the flashing tool complains if you have non-word aligned entries.

(each slot is 2 bytes)

Reply:
"On another side note, does this also consider the slot used in NSIB? So if I have NSIB+MCUBoot, it would currently be 2, but if we start supporting multi-image counters in MCUBoot, we would have 3?"

That's not how it works. It's all a bit confusing. I think we might have used the term slot too many times.

The slots we are talking about here are the slots that correspond to how many firmware updates you can do.